### ISSUE TITLE
Consistent mapping of JS async APIs to native C/C++ APIs

### ISSUE THREAD

**Issue by sbc100** [emscripten-core/emscripten#21440]
On the Web and in JS a lot APIs are async, and emscripten contains C/C++ interfaces to these APIs in several places that take several different forms.

1. APIs that take callback functions.
2. APIs that are blocking from a C/C++ POV but depend on ASYNCIFY to perform that async operation.
3. APIs return a C `promise_t` which is a reflation of a JS promise in native code.

It would be good if all APIs were available in all 3 flavors.

Additionally it would be good if APIs authors only needed to implement one of these flavors and the other two were automatically make available in terms of this single one.

TODO:

[] Decide which flavor should be the canonical one
[] Enumerate all current async APIs.  (Create a bug for each of them?)
[] Create tooling to provide different APIs flavors based on a single one
[] Move all current APIs to this new mechanism

**Comment by sbc100** [comment]
@tlively, WDYT?

**Comment by impact-maker** [comment]
Hi @sbc100 @tlively I have been researching on this project for quite some time now. After going through the documentation of asyncify (https://emscripten.org/docs/porting/asyncify.html) , I have come up with the following conclusion for this issue.

1. The second flavour (APIs that are blocking from a C/C++ POV but depend on ASYNCIFY to perform that async operation.) should be the canonical form and the api authors should only implement this option as a primary flavour and the other api functionalities should revolve around the asyncify functionality. 
2. The first functionality of callback functions of apis could be implemented with help of Sleeping / yielding feature of asyncify and the promise_t feature of C native code could use the embind/ccall of asyncify.  

I am trying to enumerate all  the asynchronous apis and will start creating issues/bugs for the same. Any feedback on the same can help me create a design/tooling for the Api migration to the new format.

**Comment by tlively** [comment]
Using the blocking calls as the canonical flavor with the other flavors built on top of it means that asyncify is always required, even if the underlying JS API is async and the C/C++ code wants to consume it as an async API. Asyncify can introduce a lot of complication and overhead, so it would be better for both users and API authors if it were necessary only for the blocking versions of APIs.

**Comment by sbc100** [comment]
> Not sure if off-topic, but there is also the em::val native C++ 20 await operator which does not require asyncify.

True, but IIUC that would not allow returning the JS event loop to process async JS events right?  So I don't think it help for async JS operations (unless I'm misunderstanding?)

**Comment by josephbirkner** [comment]
As far as I understand the example in the docs, it really should return control to the JS event loop:

![image](https://github.com/emscripten-core/emscripten/assets/8723581/fd01c579-7c62-45fa-a17b-bfdd3c5df614)

**Comment by sbc100** [comment]
Oh nice.. in that case I stand corrected.   Do you have a link to those docs?

**Comment by sbc100** [comment]
(Found it https://emscripten.org/docs/api_reference/val.h.html#_CPPv4NK10emscripten10emscripten3valcv8co_awaitEv)

**Comment by curiousdannii** [comment]
Does it require C++ all the way up the call tree? Or could a C function call that C++ function?

**Comment by tlively** [comment]
In principle C could call into C++ co_await code if it could convert the returned emscripten::val into something C can hold onto. Ideally it could be converted to an emscripten_promise_t, but I’m not sure that’s possible right now. @RReverser, do you know?

**Comment by RReverser** [comment]
> Ideally it could be converted to an emscripten_promise_t, but I’m not sure that’s possible right now. @RReverser, do you know?

No, I think that's what we discussed in that call with you and @sbc100 some time ago. `promise.h` would need to reuse `Emval.{toValue, toHandle}` to be easily convertible to `val`.

Should be a relatively straightforward change I think if someone has time.

**Comment by OmarShawky1** [comment]
Is this starter bug still available?

I know that chromium on GitHub is just a mirror so I am not sure. I have investigated the problem and will discuss details in case it is available.

**Comment by tlively** [comment]
@OmarShawky1, yes, this is still available to work on. Emscripten is a separate code base from Chromium (although we're still part of the Chromium GSOC organization), and this repo is the source of truth for Emscripten.

**Comment by impact-maker** [comment]
After going over the documentation of asyncify and basics of callbacks and promises, these are my conclusions

promise_t approach emerges as the most forward-looking choice for a canonical flavor. A tooling can then be developed to automatically generate callback-based and ASYNCIFY versions from the promise-based implementations @tlively  @sbc100

**Comment by curiousdannii** [comment]
One thing that would be nice would be if it just auto detected async functions. So instead of this:

```js
addToLibrary({
    myfunc__async: true,
    myfunc() {
        return Asyncify.handleAsync(() => {
            // Do async stuff
        })
    },
})
```

You could just write

```js
addToLibrary({
    async myfunc() {
        // Do async stuff
    },
})
```

Ideally all the various C APIs could be generated to access the one async library function.

**Comment by OmarShawky1** [comment]
# Where to start?
## My Understanding
1. API Callback is the better approach IMHO. Reasons:
  1. As aforementioned in this [comment](https://github.com/emscripten-core/emscripten/issues/21440#issuecomment-1971650855), using blocking function will block both async and sync.
  2. I am not so familiar enough with C.
  3.  I am not sure how the following would be better and how to exactly implement `promise_t` in `promise.h` as per [this comment](https://github.com/emscripten-core/emscripten/issues/21440#issuecomment-1977093912).
> APIs return a C `promise_t` which is a reflation of a JS promise in native code
 
 ## Suggested fix
 1. I should modify [emscripten_promise_await signature](https://github.com/emscripten-core/emscripten/blob/918e131fae0b5c7b1d05a5c75d7e8e676c377713/system/include/emscripten/promise.h#L147) to take an extra async function (like in [emscripten ASYNCIFY docs](https://emscripten.org/docs/api_reference/val.h.html#_CPPv4NK10emscripten10emscripten3valcv8co_awaitEv) or boolean value)
 
 So should I start editing it now or there is something that I missed?

**Comment by sbc100** [comment]
> One thing that would be nice would be if it just auto detected async functions. So instead of this:
> 
> ```js
> addToLibrary({
>     myfunc__async: true,
>     myfunc() {
>         return Asyncify.handleAsync(async () => {
>             // Do async stuff
>         })
>     },
> })
> ```
> 
> You could just write
> 
> ```js
> addToLibrary({
>     async myfunc() {
>         // Do async stuff
>     },
> })
> ```
> 
> Ideally all the various C APIs could be generated to access the one async library function.

@brendandahl @RReverser I seem to remember there was some reason why we chose not to do this already?

**Comment by RReverser** [comment]
See https://github.com/emscripten-core/emscripten/issues/20289 + https://github.com/emscripten-core/emscripten/pull/19991 for previous discussions.

**Comment by curiousdannii** [comment]
Ah, yes I can see that while a little syntax magic would be nice for simple cases like mine, if you have more complex configurations it would need a lot more magic to keep doing the right thing.
